### PR TITLE
Feature request for https://github.com/CesiumGS/cesium/issues/9545

### LINKED ISSUE #9545: Allowing user to specify getFeatureInfo URL for WMS
Hi all,

I would like to request a feature that allows users to specify the URL for the WMS getFeatureInfo.
The original ticket was here: https://github.com/TerriaJS/terriajs/issues/4751

I have a suggestion to add the feature by updating the file
under `Source/Scene/WebMapServiceImageryProvider.js`, line 185:

Original code:
`var pickFeatureResource = resource.clone();`

Suggest to change to:
```javascript
var pickFeatureResource = resource.clone();
if (defined(options.parameters.featureurl)) {
    var featureurl = Resource.createIfNeeded(options.parameters.featureurl);
    pickFeatureResource = featureurl.clone();
}
```

Thank you very much.

### PR DESCRIPTION
**Pull request by martin-bom**
Fixes #9545

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thank you so much for the pull request @martin-bom! I noticed this is your first pull request and I wanted to say welcome to the Cesium community!

The [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines) is a handy reference for making sure your PR gets accepted quickly, so make sure to skim that.

* :x: Missing `CONTRIBUTORS.md` entry.
    * Please add yourself to the [contributors](https://github.com/martin-bom/cesium/blob/wms-getfeatureurl/CONTRIBUTORS.md) file!
* :x: Missing CLA.
    * Please send in a [Contributor License Agreement](https://github.com/CesiumGS/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla) (CLA) and comment back here to let us know to check this!
    * Maintainers, [here's how to check](https://github.com/CesiumGS/cesium/blob/master/Documentation/Contributors/CLAs/README.md) for a submitted CLA.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/wms-getfeatureurl/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by ebogo1** [comment]
Hi @martin-bom - thanks for opening the PR! Could you please add a test in `WebMapServiceImageryProviderSpec` to make sure this works as expected?

**Comment by martin-bom** [comment]
> Hi @martin-bom - thanks for opening the PR! Could you please add a test in `WebMapServiceImageryProviderSpec` to make sure this works as expected?

Hi @ebogo1 I will add it and come back to you. Thank you.

File: Source/Scene/WebMapServiceImageryProvider.js [hunk 1]
```diff
@@ -184,6 +184,13 @@ function WebMapServiceImageryProvider(options) {
 
   var pickFeatureResource = resource.clone();
 
+  // If the getFeatureInfo url is specified in the parameters (featureurl),
+  // we should use it instead of the url
+  if (defined(options.parameters) && defined(options.parameters.featureurl)) {
```
Comment (ebogo1) [inline Source/Scene/WebMapServiceImageryProvider.js:189]: @tfili Could I get your thoughts on this change? I'm not sure if `options.parameters` is the right spot for the `getFeatureInfo` url here, or if it makes sense to always replace `options.url` if it exists.

Comment (tfili) [reply Source/Scene/WebMapServiceImageryProvider.js:189]: It definitely shouldn't go in `parameters`. I would do something like the following using `options.getFeatureInfoUrl`. (`options.getFeatureInfoParameters` exists so we should follow a similar naming format)

```
  var resource = Resource.createIfNeeded(options.url);

  var pickFeatureResource = defined(options.getFeatureInfoUrl) ? Resource.createIfNeeded(options.getFeatureInfoUrl) : resource.clone();
```

Comment (martin-bom) [reply Source/Scene/WebMapServiceImageryProvider.js:189]: @tfili I will update the code again based on your idea. Thank you.

Comment (martin-bom) [reply Source/Scene/WebMapServiceImageryProvider.js]: Hi @ebogo1 and @tfili I have updated the code based on your suggestions and the test cases for your review. Thank you.

File: Source/Scene/WebMapServiceImageryProvider.js [hunk 2]
```diff
@@ -268,6 +274,9 @@ function WebMapServiceImageryProvider(options) {
   pickFeatureResource.setQueryParameters(pickFeatureParams, true);
 
   this._resource = resource;
+
+  // TODO: change this pickFeatureResource to something like options.getFeaturesUrl if this option exists, else use the getCapabilities default URL
```
Comment (ebogo1) [inline Source/Scene/WebMapServiceImageryProvider.js:278]: Is this TODO comment still relevant?

File: Source/Scene/WebMapServiceImageryProvider.js [hunk 3]
```diff
@@ -268,6 +274,9 @@ function WebMapServiceImageryProvider(options) {
...
+  // UPDATE: a new paramter getFeatureInfoUrl has been introduced, if the parameter is found in options, the pickFeaturesRource will refer to the paramter
```
Comment (ebogo1) [inline Source/Scene/WebMapServiceImageryProvider.js:279]: I don't think this UPDATE comment is needed.

**Review by ebogo1 — COMMENTED** [review]
@martin-bom Apologies for the delay, some more feedback:

- There should be an entry in the `WebMapServiceImageryProvider` constructor for the new `getFeatureInfoUrl` option.
- CHANGES.md should be updated with a note that this new option was added to `WebMapServiceImageryProvider`.
- Please add your name to CONTRIBUTORS.md :) (see the concierge's comment above for details)

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @martin-bom! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

File: Source/Scene/WebMapServiceImageryProvider.js (hunk 2 above)
Comment (martin-bom) [reply Source/Scene/WebMapServiceImageryProvider.js:278]: Thank you. I will remove the comment in the next update.

File: Source/Scene/WebMapServiceImageryProvider.js (hunk 3 above)
Comment (martin-bom) [reply Source/Scene/WebMapServiceImageryProvider.js:279]: Thank you @ebogo1 , I will remove the comment as well.

**Comment by martin-bom** [comment]
Hi @ebogo1 I tried on my local machine with the fixes and they were okay. But the Travis CI Pull Request failed and there were many messages like this:
**[TSD-JSDoc] EncodedCartesian3.js:33:2 Failed to find parent of doclet 'EncodedCartesian3#low' using memberof 'EncodedCartesian3', this is likely due to invalid JSDoc.**

Is it because my PR branch is out-of-date? I was sure that I ran git pull before the commit. Thank you.

**Comment by martin-bom** [comment]
@ebogo1 Hello, I have made the changes based on your last review. Can you help to review again? Thank you.

File: Source/Scene/WebMapServiceImageryProvider.js [hunk 4]
```diff
@@ -268,6 +273,7 @@ function WebMapServiceImageryProvider(options) {
   pickFeatureResource.setQueryParameters(pickFeatureParams, true);
 
   this._resource = resource;
+
```
Comment (ebogo1) [inline Source/Scene/WebMapServiceImageryProvider.js:276]: Unneeded extra line here

**Review by ebogo1 — COMMENTED** [review]
Thanks @martin-bom - the new changes look good to me. I wonder if `getFeatureInfoUrl` is a better name than `featureInfoUrl`, but otherwise I think this is good to go.

@tfili - could you please take another look when you get a chance? Thanks!

**Comment by martin-bom** [comment]
> Thanks @martin-bom - the new changes look good to me. I wonder if `getFeatureInfoUrl` is a better name than `featureInfoUrl`, but otherwise I think this is good to go.
> 
> @tfili - could you please take another look when you get a chance? Thanks!

Hi @ebogo1 thank you for the review. I took away the "get" because when I came to the get/set definition and thought `getGetFeatureInfoUrl` was a bit awkward. I am more than happy to rename it back to `getFeatureInfoUrl` as `getFeatureInfo` is a service request name. Thank you.

File: Source/Scene/WebMapServiceImageryProvider.js (hunk 4 above)
Comment (martin-bom) [reply Source/Scene/WebMapServiceImageryProvider.js:275]: Hi @ebogo1 , I removed the extra line in this commit: https://github.com/CesiumGS/cesium/pull/9563/commits/addacfe22e476625765ca9af5ac68cdcb86b033e
Thank you.

**Pull request merged**